README.md
# pss-adobe-target

This module is a wrapper around the Adobe Target Node SDK. The module exports two constructors for use by products. One creates an express router with routes defined to get Adobe Target activities and to send notifications to Adobe Target. The other contructor creates a service that can be used to get activities or send notifications in a couple of ways, with middleware or calling  functions directly.

![Alt text](./assets/img/pss-adobe-target.svg)

## Installation

In the server directory of a product run:

```npm i -P @pss/pss-adobe-target```

## Configuration

Both the pss-adobe-target Router and Service expect the follow configuration at initialization.

```javascript
const pssAdobeTargetRouter = PssAdobeTarget.Router(adobeTargetConfig, PssLogger);
const pssAdobeTargetService = PssAdobeTarget.Service(adobeTargetConfig, PssLogger);
```

### pss-adobe-target configuration

```json
{
  "client": "vault_adobe_target_client",
  "organizationId": "vault_adobe_target_org_id",
  "propertyToken": "vault_adobe_target_property",
  "decisioningMethod": "on-device",
  "events": {
    "clientReady": onTargetClientReady,
    "artifactDownloadSucceeded": onArtifactDownloadSucceeded,
    "artifactDownloadFailed": onArtifactDownloadFailed,
  },
  "timeout": 500
}
```


For more information about the Adobe Target SDK configuration see this website: <https://adobetarget-sdks.gitbook.io/docs/sdk-reference-guides/nodejs-sdk/initialize-sdk>.

| attribute | required | type | expected value | secret | description |
| --------- | -------- | ---- | --------------- | ------ | ----------- |
| `client` | yes | string | `vault_adobe_target_client` | yes | The Adobe Target client code. This value will be replaced with a secret from vault. |
| `organizationId` | yes | string | `vault_adobe_target_org_id` | yes | The Adobe Tartet IMS Id. The value will be replaced with a secret from vault | 
| `propertyToken` | yes | string | `vault_adobe_target_property` | yes | The Adobe Target environment property. The value will be replaced with a secret from vault. | 
| `decisioningMethod` | no | string | `on-device`, `server-side`, or `hybrid` | no | Determines the decisioning method to use | 
| `events` | no | object | see below | no | An optional object with event name keys and callback function values |
| `timeout` | no | number | Timeout in milleseconds | no | The timeout value for all requests to Adobe Target. | 

### Events

These event functions will be defined either in code before initializing pss-adobe-target or in the configuration file. For more information about the Adobe Target SDK events see this website: <https://adobetarget-sdks.gitbook.io/docs/sdk-reference-guides/nodejs-sdk/sdk-events>.

| attribute | required | type | description |
| --------- | -------- | ---- | ----------- |
| `clientReady` | no | function | Emitted when the artifact has downloaded and the SDK is ready for getOffers calls. Recommended when using on-device decisioning method. |
| `artifactDownloadSucceeded` | no | function | Emitted each time a new artifact is downloaded. |
| `artifactDownloadFailed` | no | function | Emitted each time an artifact fails to download. |


### PssLogger

This is an initialized PssLogger that will be passed into both the Router and Service.

```javascript
const PssLogger = require("@pss/pss-logger")(config.loggingConfig);
```

## Exports

pss-adobe-target exports two constructors, one for an express Router and one for a Service.

### Routes

The pss-adobe-target Router sets up two routes for use in client-side ajax requests.

GET /target/activity?activity=value1&activity=value2

This route fetches one or more experience the current user is in as well as the offer delivered by those experiences. The route expects one or more activity query parameters. The activity parameter is the main mbox of the adobe target activity defined when creating the activity in Adobe Target.

PUT /target/notification?mbox=mbox1&mbox=mbox2

This route will send notifications to adobe target for one or more mboxes. The route expects one or more mbox query parameters. The mboxes correspond to the metrics setup when creating the activity in Adobe Target. This is how we will collect success metrics for server-side activities.


## Usage

### As a router

PssAdobeTarget.Router() returns an express router.

The router will mainly be used for hybrid and server-side decisioning based activities. In the main node server javascript file, usually ```app.js``` in pss products, import pss-adobe-target.

Initialize the pss adobe target router and use it.

```javascript
const app = express();

...

const PssLogger = require("@pss/pss-logger")(config.loggingConfig);
const PssAdobeTarget = require("@pss/pss-adobe-target");
...

const pssAdobeTargetRouter = PssAdobeTarget.Router(config.adobeTargetConfig, PssLogger);
app.use(pssAdobeTargetRouter);
```

#### GET /target/activity
From client-side javascript send ajax requests to fetch which experience the current user is in and the offer delivered by the experience.

The ```/target/activity``` route expects one or more ```activity=some-mbox-here``` query parameter.


```javascript
const activityName = "statdx-hsearch";

$.ajax("/target/activity?activity=" + activityName).done(function(response) {
    const statdxHsearchActivity = response[activityName];

    // Do something with the feature flag returned by the activity.
});
```

The response will be an object with a key of `experience` and a value of the experience group the user is in. This object
is setup in Adobe Target activity

Here is a sample response for the ```/target/activity?activity=statdx-hsearch``` route
```json
{
  "experience": "control"
}
```

#### PUT /target/notification

From client-side javascript send ajax requests to trigger events for currently running server-side activities.

The ```/target/notification``` route expects one or more ```mbox=some-mbox-here``` query parameter.

```javascript
const mbox = "mbox server-side-success";

$.ajax({
    url: "/target/notification?mbox=" + mbox,
    type: "PUT"
}).done(function(response) {
    // do something with the response
});
```

Here is an example of the notification route response object

```json
{
  "status": 200,
  "requestId": "7d62c85c07fc4fe3a4622a0f847903ce",
  "id": {
    "tntId": "67664e186ef54ffd9cb60b20ef66aceb.35_0"
  }
}
```

### As middleware

PssAdobeTarget.Service() returns an object with a number of functions for use interacting with adobe target. Among these function are two middleware methods that will add a function to the request to fetch activities or send notifications.

First initialize the Service

```javascript
const PssLogger = require("@pss/pss-logger")(config.loggingConfig);
const PssAdobeTarget = require("@pss/pss-adobe-target");
...

const pssAdobeTargetService = PssAdobeTarget.Service(config.adobeTargetConfig, PssLogger);
```

Then use the middleware on routes defined in the product. 

The ```Middleware.getActivities``` expects one or more ```activity=activity-box``` query paramenter.

The ```Middleware.sendNotification``` expects one or more ```mbox=mbox-name``` query paramenter.

```javascript
    app.get("/test/target/activity",
        targetService.Middleware.getActivities,
        (req, res, next) => {
            const activities = req.getTargetActivities();
            const searchEndpoint = activities["statdx-hsearch"];
            
            if (searchEndpoint.experience === "control") {
                // show the control experience
            }
            if (searchEndpoint.experience === "variation") {
                // show the variation experience
            }
            else {
                // show the default experience
            }
            
        });


    app.get("/test/target/notification",
        targetService.Middleware.sendNotification,
        (req, res, next) => {
            const activities = req.getTargetActivities();
            
            const notificationResponse = req.getTargetNotificationResponse();
            if (notificationResponse.status !== 200) {
                // Do something if the notification failed to send
            }
        });
```

### In code

The Service exports functions that can be used in nodejs code to interact with adobe target

```javascript

const targetService = PssAdobeTarget.Service(adobeTargetConfig, PssLogger);

const renderSomePage = (req, res, next) => {
    
    ...
    
    const userId = req.session.userId;
    const sessionId = req.session.authToken;
    const targetCookie = req.cookies.targetCookie ? req.cookies.targetCookie : null;
    const activities = ["show-some-feature"];
    
    targetService.getActivities(activities, userId, sessionId, targetCookie)
        .then(response => {
            logger.trace("Get activity response = ");
            logger.trace(response);

            const experience = targetService.getExperienceFromTargetResponse(response, "show-some-feature");
            
            let showNewFeature;
            if (experience === "control") {
                showNewFeature = false;
            }
            else if (experience === "variation") {
                showNewFeature = true;
            }
            else {
                showNewFeature = req.session.features["show-some-feature"].active;
                logger.error(`Failed to get activity experience for "${activities}"`);
            }

            res.render("/some/page", {
                showVariation: showNewFeature
            });
            
        })
        .catch(error => {
            logger.error("Failed to get activities");
            logger.error(error);
            res.render("/error/page");
        });
};

const sendSomeNotification = (req, res, next) => {

...

    const userId = req.session.userId;
    const sessionId = req.session.authToken;
    const targetCookie = req.cookies.targetCookie ? req.cookies.targetCookie : null;
    const mboxes = req.query.mbox || ["some-mbox", "another-mbox"];
    const eventType = req.query.eventType;

    targetService.sendNotification(mboxes, eventType, userId, sessionId, targetCookie)
        .then(response => {
            logger.trace("Send notification response = ");
            logger.trace(response);

            ...

        })
        .catch(error => {
            logger.error("Failed to get activities");
            logger.error(error);
        });
};

```



CHANGELOG.md
# Changelog
_All notable changes to this component will be documented in this file._

_The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)._

This component adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased - X.Y.Z] - YYYY-MM-DD

## [1.2.1] - 2022-04-19
### Changed
- Use userId as sessionId for all middleware, routes, and service methods

## [1.2.0] - 2022-03-18
### Added
- TargetService.getExperienceFromTargetResponse
- Linting

### Changed
- Update documentation for getActivity

## [1.1.0] - 2021-05-27
### Changed
- fixed bugs with creating adobe target API request objects.
- added sessionId and thirdPartyId to requests to adobe target.


## [1.0.0] - 2021-05-27
### Added
- Initial implementation -- see README.md

test/config/default-template.js
// copy and rename to `default.js` and replace placeholder values
module.exports = {
  adobeTargetConfig: {
    client: "[vault_adobe_target_client]",
    organizationId: "[vault_adobe_target_org_id]",
    decisioningMethod: "server-side",
    propertyToken: "[vault_adobe_target_property]"
  }
};


lib/TargetRouter.js

// ============================================================
// Requires
// ============================================================

const assert = require("assert");
const express = require("express");

const TargetService = require("./TargetService");

// ============================================================
// Variables
// ============================================================

let logger;
let targetService;

// ============================================================
// Exports
// ============================================================

exports = module.exports = targetRouterConstructor;

// ============================================================
// Functions
// ============================================================

/**
 * Returns an express router with routes for getting target activities and sending event notifications
 * @param targetConfig
 * @param PssLogger
 */
function targetRouterConstructor(targetConfig, PssLogger) {
    assert(PssLogger, "Configured PssLogger must be provided");
    assert("Logger" in PssLogger && typeof PssLogger.Logger === "function", "PssLogger must contain Logger function");
    logger = PssLogger.Logger(module.filename);

    targetService = TargetService(targetConfig, PssLogger);

    // create the express router
    let targetRouter = express.Router();

    targetRouter.get("/target/activity", targetService.Route.getActivities);
    targetRouter.put("/target/notification", targetService.Route.sendNotification);

    logger.info("Mounting pss-target at " + module.filename);
    return targetRouter;
}


lib/TargetService.js

// ============================================================
// Requires
// ============================================================

const assert = require("assert");
const util = require("util");
const { v4: uuidv4 } = require("uuid");

const TargetClient = require("@adobe/target-nodejs-sdk");

// ============================================================
// Variables
// ============================================================

let logger;
let targetClient;

// ============================================================
// Functions
// ============================================================

/**
 * Returns a configured target SDK client
 * @param Object targetConfig
 */
const initializeTargetClient = (targetConfig) => {
    validateConfig(targetConfig);

    logger.debug("Creating the target client :: config = " + util.inspect(targetConfig));
    return TargetClient.create(targetConfig);
};

// ---------- Business Logic -----------

/**
 * Pulls out a unique sessionId out of the express session.
 * @param session express session object
 * @returns string sessionId
 */
const getSessionId = (session) => {
    const sessionId = session && session.authToken ? session.authToken : uuidv4();
    return sessionId;
};

/**
 * Pulls out an unique userId out of the express session.
 * @param session express session object
 * @returns string userId
 */
const getUserId = (session) => {
    if (!session || !session.userId || !session.authToken) {
        return uuidv4();
    }

    const userId = session.isAnonymous ? session.userId + "_" + session.authToken : session.userId;
    return userId;
};

/**
 * Creates an array of mbox objects.
 * @param mboxes array of strings
 * @param eventType string
 * @returns {*[{type, timestamp, id, name, index}]}
 */
const createMboxArray = (mboxes, eventType) => {
    const mboxArray = [];

    if (!Array.isArray(mboxes)) {
        return mboxArray;
    }

    const type = eventType === "click" ? "click" : "display";
    let index = 0;
    for (const mbox of mboxes) {
        mboxArray.push({
            type: type,
            timestamp: Date.now(),
            id: uuidv4(),
            name: mbox,
            index: index
        });
        index++;
    }

    return mboxArray;
};

/**
 * Creates an adobe target request object. Currently the only required field we need is thirdPartyId.
 * https://developers.adobetarget.com/api/delivery-api/#tag/Delivery-API
 * @param string userId
 */
const createTargetRequest = (userId) => {

    const request = {
        "id": {
            "thirdPartyId": userId
        }
    };

    return request;
};

/**
 * Creates an adobe target options object.
 * https://adobetarget-sdks.gitbook.io/docs/sdk-reference-guides/nodejs-sdk/get-offers#parameters
 * @param request adobe target request object. https://developers.adobetarget.com/api/delivery-api/#tag/Delivery-API
 * @param targetCookie adobe target cookie. https://adobetarget-sdks.gitbook.io/docs/sdk-reference-guides/nodejs-sdk/get-offers#promise
 * @param sessionId string
 * @returns {{request, targetCookie, sessionId}}
 */
const createTargetOptions = (request, targetCookie, sessionId) => {
    const options = {
        request: request,
        targetCookie: targetCookie,
        sessionId: sessionId
    };
    return options;
};

/**
 * Creates an array of adobe target notification objects. https://developers.adobetarget.com/api/delivery-api/#operation/execute
 * @param mboxes array of strings
 * @param eventType string
 * @returns {*[{id, mbox, type, timestamp}]}
 */
const createNotifications = (mboxes, eventType) => {
    const notifications = [];

    if (!Array.isArray(mboxes)) {
        return notifications;
    }

    const type = eventType === "click" ? "click" : "display";

    for (const mbox of mboxes) {

        const notification = {
            id: uuidv4(),
            mbox: {
                name: mbox,
            },
            type: type,
            timestamp: Date.now()
        };

        notifications.push(notification);
    }

    return notifications;
};

/**
 * Creates an adobe target notification request
 * @param [notifications] array of adobe target notification objects
 * @param userId string
 * @returns {{request: {id, notifications: *[]}, targetCookie}}
 */
const createNotificationRequest = (notifications, userId) => {

    const request = {
        ...createTargetRequest(userId),
        "notifications": notifications
    };

    return request;
};

/**
 *
 * @param mboxes array of adobe target mbox objects. https://developers.adobetarget.com/api/delivery-api/#operation/execute
 * @param userId string
 * @returns {{id: {thirdPartyId: *}, execute: {mboxes}}}
 */
const createOfferRequest = (mboxes, userId) => {
    logger.debug("TargetService.createOfferRequest");

    const request = {
        ...createTargetRequest(userId),
        "execute": {
            "mboxes": mboxes
        }
    };
    return request;
};

/**
 * Asserts that the passed in config object contains all required elements
 * @param Object targetConfig
 */
const validateConfig = (targetConfig) => {

    logger.trace("TargetService.validateConfig :: config = " + util.inspect(targetConfig));

    assert(targetConfig, "adobe target sdk config must be provided");
    assert(targetConfig instanceof Object, "adobe target sdk config must be an object");

    assert(targetConfig.client, "targetConfig.client must be provided");
    assert((typeof targetConfig.client === "string"), "targetConfig.client must be a string");

    assert(targetConfig.organizationId, "targetConfig.organizationId must be provided");
    assert((typeof targetConfig.organizationId === "string"), "targetConfig.organizationId must be a string");

    assert(targetConfig.propertyToken, "targetConfig.propertyToken must be provided");
    assert((typeof targetConfig.propertyToken === "string"), "targetConfig.propertyToken must be a string");
};

/**
 * Extracts the experience from the activity offer defined in adobe target. The value will be 'control' or 'variation'.
 * If the experiment isn't running or there is an error, undefined will be returned and an error will be logged.
 * @param response  see {$link https://developers.adobetarget.com/api/delivery-api/#section/Single-or-Batch-Delivery/Batched-Delivery-Call} for documentation on AT getOffers response.
 * @param mboxName string - This is the main mbox location defined in Adobe Target
 * @returns String experience {'control', 'variation', or undefined}
 */
const getExperienceFromTargetResponse = (response, mboxName) => {
    let experience;

    try {
        // Activity response contains a response object. Gracefully handle either
        const res = response.response ? response.response : response;

        const mboxes = res.execute.mboxes;

        for (const mbox of mboxes) {
            if (mbox.name === mboxName &&
                mbox.options &&
                mbox.options.length > 0 &&
                mbox.options[0].content
            ) {
                // This assumes the Adobe Target offer was created with the best practice pattern defined in the README. { "experience": variation_name_here }
                // We should only create a single offer per experience.
                experience = mbox.options[0].content.experience;
            }
        }
    }
    catch (error) {
        logger.error("Unexpected Adobe Target activity response received for mbox = " + mboxName);
        logger.error("response = " + response);
        logger.error(error);
    }

    return experience;
};

// ---------- Promises -----------

// TODO - Add getExperience https://elsevier.atlassian.net/browse/AMP-4737

/**
 * Returns the promise created by adobe target sdk getAttributes function.
 * @param activities array of strings
 * @param userId string
 * @param sessionId string
 * @param targetCookie adobe target cookie. https://adobetarget-sdks.gitbook.io/docs/sdk-reference-guides/nodejs-sdk/get-offers#promise
 * @returns {*|(function(): Promise<*>)}
 */
const getActivities = (activities, userId, sessionId, targetCookie) => {

    const mboxes = createMboxArray(activities, "display");
    const request = createOfferRequest(mboxes, userId);
    const options = createTargetOptions(request, targetCookie, sessionId);

    return targetClient.getOffers(options);
};

/**
 * Returns the promise created by adobe target sdk sendNotifications function.
 * @param mboxes array of strings
 * @param eventType string (display | click)
 * @param userId string
 * @param sessionId string
 * @param targetCookie adobe target cookie. https://adobetarget-sdks.gitbook.io/docs/sdk-reference-guides/nodejs-sdk/get-offers#promise
 * @returns {*|(function(): Promise<*>)}
 */
const sendNotification = (mboxes, eventType, userId, sessionId, targetCookie) => {
    const notifications = createNotifications(mboxes, eventType);
    const request = createNotificationRequest(notifications, userId);
    const options = createTargetOptions(request, targetCookie, sessionId);

    return targetClient.sendNotifications(options);
};

// ---------- Middleware Routes -----------

/**
 * Adds a getTargetActivites function to req object for use by subsequent middleware / route definitions
 * @param req
 * @param res
 * @param next
 */
const getActivitiesMiddleware = (req, res, next) => {
    logger.debug("TargetService.getActivitiesMiddleware");

    const activity = req.query.activity;
    const activities = Array.isArray(activity) ? [...activity] : [activity];

    const userId = getUserId(req.session);
    const sessionId = getUserId(req.session);
    const targetCookie = req.cookies.targetCookie ? req.cookies.targetCookie : null;

    getActivities(activities, userId, sessionId, targetCookie)
        .then((response) => {
            logger.debug("Target activities :: " + util.inspect(response.asObject()));

            // Add getTargetActivities function to the request.
            req.getTargetActivities = () => {
                return response.asObject();
            };

            next();
        })
        .catch((error) => {
            logger.error("Error fetching adobe target activities :: activities = " + util.inspect(activities));
            logger.error(error);

            req.getTargetActivities = () => {
                return {
                    "error": {
                        status: 500,
                        message: "Error fetching adobe target activities :: activities = " + util.inspect(activities),
                        error: error
                    }
                };
            };

            next();
        });
};

/**
 * Adds a getTargetNotificationResponse function to req object for use by subsequent middleware / route definitions
 * @param req
 * @param res
 * @param next
 */
const sendNotificationMiddleware = (req, res, next) => {
    logger.debug("TargetService.sendNotificationsMiddleware");

    const mbox = req.query.mbox;
    const eventType = req.query.type || "display";
    const mboxes = Array.isArray(mbox) ? [...mbox] : [mbox];

    const userId = getUserId(req.session);
    const sessionId = getUserId(req.session);
    const targetCookie = req.cookies.targetCookie ? req.cookies.targetCookie : null;

    sendNotification(mboxes, eventType, userId, sessionId, targetCookie)
        .then((response) => {
            logger.debug("Notifications sent to adobe target :: response " + util.inspect(response));

            // Add getTargetNotificationResponse function to the request.
            req.getTargetNotificationResponse = () => {
                let targetResponse = {...response.response};
                delete targetResponse.client;
                delete targetResponse.edgeHost;
                return targetResponse;
            };

            next();
        })
        .catch((error) => {
            logger.error("Error sending notifications to adobe target :: mboxes = " + util.inspect(mboxes));
            logger.error(error);

            req.getTargetNotificationResponse = () => {
                return {
                    status: 500,
                    message: "Error sending notifications to adobe target :: mboxes = " + util.inspect(mboxes),
                    error: error
                };
            };

            next();
        });
};

// ---------- Routes -----------

/**
 * Requests activities from adobe target. One or more "activity" query parameters expected.
 * @param req
 * @param res
 * @param next
 */
const getActivitiesRoute = (req, res) => {
    logger.debug("TargetRouter.getActivity");

    const activity = req.query.activity;
    const activities = Array.isArray(activity) ? [...activity] : [activity];

    const userId = getUserId(req.session);
    const sessionId = getUserId(req.session);
    const targetCookie = req.cookies.targetCookie ? req.cookies.targetCookie : null;

    getActivities(activities, userId, sessionId, targetCookie)
        .then((response) => {
            logger.debug("Target activities :: " + util.inspect(response));
            res.json(response).status(200);
        })
        .catch((error) => {
            logger.error("Error fetching adobe target activity [" + activity + "]");
            logger.error(error);
            res.status(500).json({
                message: "Error fetching adobe target activity [" + activity + "]",
                error: error
            });
        });
};

/**
 * Sends notifications to adobe target. One or more "mbox" query parameters expected.
 * @param req
 * @param res
 * @param next
 */
const sendNotificationRoute = (req, res) => {
    logger.debug("TargetRouter.sendNotification");

    const mbox = req.query.mbox;
    const eventType = req.query.eventType || "display";
    const mboxes = Array.isArray(mbox) ? [...mbox] : [mbox];

    const userId = getUserId(req.session);
    const sessionId = getUserId(req.session);
    const targetCookie = req.cookies.targetCookie ? req.cookies.targetCookie : null;

    sendNotification(mboxes, eventType, userId, sessionId, targetCookie)
        .then((response) => {
            logger.debug("Notifications sent to adobe target :: response " + util.inspect(response));
            let targetResponse = {...response.response};
            delete targetResponse.client;
            delete targetResponse.edgeHost;
            res.json(targetResponse).status(200);
        })
        .catch((error) => {
            logger.error("Error sending notifications to adobe target :: mboxes = " + util.inspect(mboxes));
            logger.error(error);
            res.status(500).json({
                message: "Error sending notification for mbox [" + mbox + "]",
                error: error
            });
        });

};

// ============================================================
// Constructor
// ============================================================

/**
 * Returns a TargetService object.
 * @param targetConfig
 * @param PssLogger
 * @returns {TargetService}
 */
const targetServiceConstructor = (targetConfig, PssLogger) => {
    assert(PssLogger, "Configured PssLogger must be provided");
    assert("Logger" in PssLogger && typeof PssLogger.Logger === "function", "PssLogger must contain Logger function");
    logger = PssLogger.Logger(module.filename);

    targetClient = initializeTargetClient(targetConfig);

    let targetService = {
        Route: {},
        Middleware: {}
    };

    // Export route functions
    targetService.Route.getActivities = getActivitiesRoute;
    targetService.Route.sendNotification = sendNotificationRoute;

    // Export middleware functions
    targetService.Middleware.getActivities = getActivitiesMiddleware;
    targetService.Middleware.sendNotification = sendNotificationMiddleware;

    // Export business logic functions
    targetService.createMboxArray = createMboxArray;
    targetService.createNotificationRequest = createNotificationRequest;
    targetService.createNotifications = createNotifications;
    targetService.createOfferRequest = createOfferRequest;
    targetService.createTargetOptions = createTargetOptions;
    targetService.createTargetRequest = createTargetRequest;
    targetService.getActivities = getActivities;
    targetService.getSessionId = getSessionId;
    targetService.getUserId = getUserId;
    targetService.sendNotification = sendNotification;
    targetService.validateConfig = validateConfig;
    targetService.getExperienceFromTargetResponse = getExperienceFromTargetResponse;

    return targetService;
};

// ============================================================
// Exports
// ============================================================

exports = module.exports = targetServiceConstructor;

exports.initializeTargetClient = initializeTargetClient;
